test/controllers/products_controller_test.rb

```diff
@@ -2,28 +2,145 @@
+      must_respond_with :success
+    end
+
+    it "should give a flash notice instead of showing a non-existant, invalid product" do
+      product = existing_product
+      invalid_product_id = product.id
+      product.destroy
+
+      get product_path(invalid_product_id)
+
+      must_respond_with :not_found
+      expect(flash[:error]).must_equal "Could not find product with id: #{invalid_product_id}"
+    end
+  end
+
+  # describe "new" do
+  #   it "succeeds" do
+  #     get new_product_path
+
+  #     must_respond_with :success
+  #   end
   # end
 
+  describe "create" do
+    describe "logged in users" do
+      before do
+        perform_login(users(:dee))
+        merchant = new_merchant
+      end
+
+      it "will save a new product and redirect if given valid inputs" do
+        input_photo_url = "https://drive.google.com/uc?id=1LCGn0419g0STAeyDo-miWCD6o5ZOEkXM"
+        input_description = "black, lightweight, breathable wool turtleneck, perfect for those moments when you're sweating your scam."
+        input_name = "Iconic Tech Turtleneck"
+        input_price = 9900
+        input_quantity = 20
+        input_retired = false
+        input_user_id = new_merchant.id
+        test_input = {
+          "product": {
+            photo_url: input_photo_url,
+            description: input_description,
+            name: input_name,
+            price: input_price,
+            quantity: input_quantity,
+            retired: input_retired,
+            user_id: input_user_id,
+          },
+        }
+
+        expect {
+          post products_path, params: test_input
+        }.must_change "Product.count", 1
+
+        new_product = Product.find_by(name: input_name)
+        expect(new_product).wont_be_nil
+        expect(new_product.name).must_equal input_name
+        expect(new_product.price).must_equal input_price
+        expect(new_product.description).must_equal input_description
+
+        must_respond_with :redirect
+      end
+      it "will give a 400 error with invalid params" do
+        input_photo_url = "https://drive.google.com/uc?id=1LCGn0419g0STAeyDo-miWCD6o5ZOEkXM"
+        input_description = "black, lightweight, breathable wool turtleneck, perfect for those moments when you're sweating your scam."
+        input_name = ""
+        input_price = 9900
+        input_quantity = 20
+        input_retired = false
+        input_user_id = new_merchant.id
+        test_input = {
+          "product": {
+            photo_url: input_photo_url,
+            description: input_description,
+            name: input_name,
+            price: input_price,
+            quantity: input_quantity,
+            retired: input_retired,
+            user_id: input_user_id,
+          },
+        }
+        expect {
+          post products_path, params: test_input
+        }.wont_change "Product.count"
+
+        expect(flash[:name]).must_equal ["can't be blank"]
+        must_respond_with :bad_request
```

Comment:
I think we'd also want to test for our flash message here.  Something like this:
expect(flash[:error]).must_equal "Save was unsuccessful. Try again!"

Reply:
i throw a general error here because it's the same conditional statement for both invalid info and users who haven't logged in - i did just add another test that checks the error message for a guest user trying to create a product though

Reply:
ahhh ok!  that totes makes sense.